### PR TITLE
single shooting solver

### PR DESCRIPTION
**Pull request by ag2718**
## Checklist

- [ ] Appropriate tests were added
- [x] Any code changes were done in a way that does not break public API
- [ ] All documentation related to code changes were updated
- [x] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [x] Any new documentation only uses public API
  
## Additional context

Implementation of basic single shooting, will add batching / multi-shooting / more testing soon.

### CODE REVIEW

File: src/SimpleBoundaryValueDiffEq.jl [hunk 1]
```diff
@@ -1,5 +1,19 @@
...
+    function ivp_deviation(u0, p)
+        ivp = ODEProblem(prob.f, u0, prob.tspan, 0)
+        ivp_sol = solve(ivp)
+        prob.f.bc(ivp_sol, 0, 0)
+    end
+
+    root_finding_prob = NonlinearProblem(ivp_deviation, prob.u0, 0)
+    best_u0 = solve(root_finding_prob)
+    ivp_with_init_condition = ODEProblem(prob.f, best_u0, prob.tspan)
+    solve(ivp_with_init_condition)
```
Comment (avik-pal) [inline src/SimpleBoundaryValueDiffEq.jl:16]: Try to see how BoundaryValueDiffEq is setup with CommonSolve.jl

You would want to specify this as `SimpleSingleShooting(nlsolve_algorithm)`

Then you call `solve(prob, alg)`

File: src/SimpleBoundaryValueDiffEq.jl [hunk 2]
```diff
@@ -1,5 +1,19 @@
...
 
-# Write your package code here.
+using DiffEqBase, DifferentialEquations, NonlinearSolve, Plots
 
+# Only for out-of-place BVP, in place needs to be converted to out-of-place
+function single_shooting(prob, p)
+    function ivp_deviation(u0, p)
+        ivp = ODEProblem(prob.f, u0, prob.tspan, 0)
+        ivp_sol = solve(ivp)
+        prob.f.bc(ivp_sol, 0, 0)
```
Comment (avik-pal) [inline src/SimpleBoundaryValueDiffEq.jl:10]: not all systems take in the ivp_sol that is only true for TP BVP

File: src/SimpleBoundaryValueDiffEq.jl [hunk 3]
```diff
@@ -1,5 +1,19 @@
 module SimpleBoundaryValueDiffEq
 
-# Write your package code here.
+using DiffEqBase, DifferentialEquations, NonlinearSolve, Plots
 
+# Only for out-of-place BVP, in place needs to be converted to out-of-place
```
Comment (avik-pal) [inline src/SimpleBoundaryValueDiffEq.jl:5]: No don't worry about inplace, converting inplace to an OOP is almost always a terrible idea.

File: test/runtests.jl [hunk 4]
```diff
@@ -1,4 +1,4 @@
-using SimpleBoundaryValueDiffEq
+using .SimpleBoundaryValueDiffEq
```
Comment (avik-pal) [inline test/runtests.jl:1]: You shouldn't need this

File: src/BVPTest.jl [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+function test_bc(u, p, t)
+    [u[1][1] - 1, u[end][1]]
+end
+
+prob = BVProblem(test_f, test_bc, [0.0, 0.0], (0.0, 1.0))
+final_sol = SimpleBoundaryValueDiffEq.single_shooting(prob, 0.0)
+
+plot(final_sol, linewidth = 2, title = "Solution of the ODE",
+    xaxis = "Time (t)", yaxis = "u(t)", label = "u(t)")
```
Comment (avik-pal) [inline src/BVPTest.jl:17]: See how Julia testing is setup. See SimpleNonlinearSolve.jl as an example

**Review by avik-pal — CHANGES_REQUESTED** [review]

File: src/SimpleBoundaryValueDiffEq.jl [hunk 6]
```diff
@@ -1,5 +1,19 @@
 module SimpleBoundaryValueDiffEq
 
-# Write your package code here.
+using DiffEqBase, DifferentialEquations, NonlinearSolve, Plots
```
Comment (ErikQQY) [inline src/SimpleBoundaryValueDiffEq.jl:3]: As for the nonlinear solving in simple shooting methods, I think SimpleNonlinearSolve.jl is enough?

Comment (avik-pal) [reply src/SimpleBoundaryValueDiffEq.jl:3]: Yes we should just use simplenonlinearsolve

**Pull request closed without merging**